scylla/src/transport/downgrading_consistency_retry_policy.rs

```diff
@@ -165,3 +165,466 @@ impl RetrySession for DowngradingConsistencyRetrySession {
         *self = DowngradingConsistencyRetrySession::new();
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::{io::ErrorKind, sync::Arc};
+
+    use bytes::Bytes;
+    use scylla_cql::errors::BadQuery;
+
+    use super::*;
+    use crate::retry_policy::DefaultRetryPolicy;
+
+    const CONSISTENCY_LEVELS: &[Consistency] = &[
+        Consistency::All,
+        Consistency::Any,
+        Consistency::EachQuorum,
+        Consistency::LocalOne,
+        Consistency::LocalQuorum,
+        Consistency::One,
+        Consistency::Quorum,
+        Consistency::Three,
+        Consistency::Two,
+    ];
+
+    fn make_query_info_with_cl(
+        error: &QueryError,
+        is_idempotent: bool,
+        cl: Consistency,
+    ) -> QueryInfo<'_> {
+        QueryInfo {
+            error,
+            is_idempotent,
+            consistency: LegacyConsistency::Regular(cl),
+        }
+    }
+
+    // Asserts that downgrading consistency policy never retries for this Error
+    fn downgrading_consistency_policy_assert_never_retries(error: QueryError, cl: Consistency) {
+        let mut policy = DowngradingConsistencyRetryPolicy::new().new_session();
+        assert_eq!(
+            policy.decide_should_retry(make_query_info_with_cl(&error, false, cl)),
+            RetryDecision::DontRetry
+        );
+
+        let mut policy = DefaultRetryPolicy::new().new_session();
+        assert_eq!(
+            policy.decide_should_retry(make_query_info_with_cl(&error, true, cl)),
+            RetryDecision::DontRetry
+        );
+    }
+
+    #[test]
+    fn downgrading_consistency_never_retries() {
+        let never_retried_dberrors = vec![
+            DbError::SyntaxError,
+            DbError::Invalid,
+            DbError::AlreadyExists {
+                keyspace: String::new(),
+                table: String::new(),
+            },
+            DbError::FunctionFailure {
+                keyspace: String::new(),
+                function: String::new(),
+                arg_types: vec![],
+            },
+            DbError::AuthenticationError,
+            DbError::Unauthorized,
+            DbError::ConfigError,
+            DbError::ReadFailure {
+                consistency: LegacyConsistency::Regular(Consistency::Two),
+                received: 2,
+                required: 1,
+                numfailures: 1,
```

Comment:
Is this example realistic? The coordinator received two responses, including one error - so the other was a success, right? If one succeeded and one was required for consistency, then why did it fail? (I may be missing something obvious here).

Reply:
Any suggestions how I could improve this test?

Reply:
Is it as simple as swapping values of `received` with `required` ?